sudoku.py

```diff
@@ -0,0 +1,34 @@
+def is_valid(num, row, col, puzzle):
+    if num in puzzle[row]:
+        return False
+
+    for i in range(9):
+        if puzzle[i][col] == num:
+            return False
+
+    rowstart = 3 * (row // 3)
+    colstart = 3 * (col // 3)
+
+    for r in range(rowstart, rowstart + 3):
+        for c in range(colstart, colstart + 3):
+            if puzzle[r][c] == num:
```

Comment:
usually single letter variable names are frowned upon, but in this case it's pretty clear how they are being used.